pkg/lib/catalogsource/image_template.go

```diff
@@ -0,0 +1,356 @@
+func convertToKey(key GVK_Key, jsonPath string) string {
+	return fmt.Sprintf("{group:%s,version:%s,kind:%s,name:%s,namespace:%s,jsonpath:%s}", key.Group, key.Version, key.Kind, key.name, key.namespace, jsonPath)
+}
+
+// gvkToJSONPathMap is a multimap (i.e. one key with multiple values) where each value is
+// zero or more JSON paths. In other words the user could specify multiple JSON path references
+// for the same kubernetes manifest
+var gvkToJSONPathMap = map[GVK_Key][]string{}
+
+func init() {
+	// Handle known static templates
+	initializeIfNeeded(TEMPL_KUBEMAJORV)
+	initializeIfNeeded(TEMPL_KUBEMINORV)
+	initializeIfNeeded(TEMPL_KUBEPATCHV)
+}
+
+// initializeIfNeeded sets the map to a "missing" value if its not already present
+func initializeIfNeeded(templateKey string) {
+	// have we encountered this template already?
+	if _, ok := templateNameToReplacementValuesMap[templateKey]; !ok {
+		// this is a new template, so default to missing value
+		templateNameToReplacementValuesMap[templateKey] = templateMissing
+	}
+}
+
+// resetMaps is only useful for test cases
+func resetMaps() {
+	templateNameToReplacementValuesMap = map[string]string{}
+	gvkToJSONPathMap = map[GVK_Key][]string{}
+	initializeIfNeeded(TEMPL_KUBEMAJORV)
+	initializeIfNeeded(TEMPL_KUBEMINORV)
+	initializeIfNeeded(TEMPL_KUBEPATCHV)
+}
+
+type regexEntry struct {
+	captureGroup string
+	template     string
+}
+
+func (r *regexEntry) String() string {
+	return fmt.Sprintf(`(?P<%s>%s)`, r.captureGroup, r.template)
+}
+
+type regexEntries []regexEntry
+
+func (r regexEntries) String() string {
+	regexEntryAsString := []string{}
+	for _, regexEntry := range r {
+		regexEntryAsString = append(regexEntryAsString, regexEntry.String())
+	}
+	result := strings.Join(regexEntryAsString, "|")
+	return result
+}
+
+var regexList = regexEntries{
+	{cap_grp_kubemajorv, TEMPL_KUBEMAJORV},
+	{cap_grp_kubeminorv, TEMPL_KUBEMINORV},
+	{cap_grp_kubepatchv, TEMPL_KUBEPATCHV},
+	{cap_grp_gvk, templ_gvk},
+}
+
+var regexImageTemplates = regexp.MustCompile(regexList.String())
+
+var regexGVKTemplate = regexp.MustCompile(templ_gvk)
+
+// ReplaceTemplates takes a catalog image reference containing templates (i.e. catalogImageTemplate)
+// and attempts to replace the templates with actual values (if available).
+// The return value processedCatalogImageTemplate represents the catalog image reference after processing.
+// Callers of this function should check the unresolvedTemplates return value to determine
+// if all values were properly resolved (i.e. empty slice means all items were resolved, and presence
+// of a value in the slice means that the template was either not found in the cache or its value has not been
+// fetched yet). Providing an empty catalogImageTemplate results in empty processedCatalogImageTemplate and
+// zero length unresolvedTemplates
+func ReplaceTemplates(catalogImageTemplate string) (processedCatalogImageTemplate string, unresolvedTemplates []string) {
```

Comment:
without the dynamic gvk templating, this could probably be simplified to a call to the stdlib [template package](https://pkg.go.dev/text/template)

Reply:
also: since any unresolved components should mean the image reference is invalid, I'm not sure I get the value in returning a partially-resolved reference and a list of the unresolved ones: why not just return an error with the missing values?

Reply:
The unresolvedTemplates are used to inform the user in a message about exactly what template that was not resolvable. It is used in messages exposed in both the log and the status conditions.

Reply:
Also the processedCatalogImageTemplate is used to inform the user of what the template looks like in the status conditions. It helps the user see what's wrong by looking at the message AND what the templating code computed.

Reply:
I'd like to address the security shortcomings of the GVK feature in the future, so I'd like to keep things as close to "working" as possible.